dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -272,13 +274,118 @@ Int64 WindowTransformAction::getPartitionEndRow(size_t block_rows)
     return left;
 }
 
-void WindowTransformAction::advanceFrameStart()
+RowNumber WindowTransformAction::stepForward(const RowNumber & current_row, Int64 n)
 {
-    if (frame_started)
+    auto dist = distance(current_row, partition_start);
+    assert(dist >= 0);
+    if (dist <= n)
+        return partition_start;
+
+    RowNumber result_row = current_row;
+
+    // The step happens only in a block
+    if (static_cast<Int64>(result_row.row) >= n)
     {
-        return;
+        result_row.row -= n;
+        return result_row;
+    }
+
+    // The step happens between blocks
+    n -= result_row.row + 1;
+    --result_row.block;
+    result_row.row = blockAt(result_row).rows - 1;
+    while (n > 0)
+    {
+        auto & block = blockAt(result_row);
+        if (static_cast<Int64>(block.rows) > n)
+        {
+            result_row.row = block.rows - n - 1; // index, so we need to -1
+            break;
+        }
+        n -= block.rows;
+        --result_row.block;
+        result_row.row = blockAt(result_row).rows - 1;
+    }
+    return result_row;
+}
+
+std::tuple<RowNumber, bool> WindowTransformAction::stepBackward(const RowNumber & current_row, Int64 n)
+{
+    // Range of rows is [frame_start, frame_end),
+    // and frame_end position is behind the position of the last frame row.
+    // So we need to ++n.
+    ++n;
+
+    auto dist = distance(partition_end, current_row);
+    assert(dist >= 1);
+
+    // Distance is too long and partition_end is the longest distance.
```

Comment:
`Distance is too long` makes me a little confused

Reply:
> `Distance is too long` makes me a little confused

`Offset is too large and the partition_end is the longest position we can reach` how about this comment?